linkerd/proxy/tcp/src/balance.rs

```diff
@@ -1,28 +1,86 @@
 use linkerd_error::Error;
-use linkerd_stack::layer;
+use linkerd_proxy_core::Resolve;
+use linkerd_proxy_discover::{self as discover, NewSpawnDiscover};
+use linkerd_stack::{layer, NewService, Param, Service};
 use rand::thread_rng;
-use std::{hash::Hash, time::Duration};
-pub use tower::{
-    balance::p2c::Balance,
-    load::{Load, PeakEwmaDiscover},
+use std::{marker::PhantomData, net::SocketAddr, time::Duration};
+use tower::{
+    balance::p2c,
+    load::{CompleteOnResponse, PeakEwmaDiscover},
 };
-use tower::{discover::Discover, load::CompleteOnResponse};
-
-/// Produces a PeakEWMA balancer that uses connect latency (and pending
-/// connections) as its load metric.
-pub fn layer<T, D>(
-    default_rtt: Duration,
-    decay: Duration,
-) -> impl tower::layer::Layer<D, Service = Balance<PeakEwmaDiscover<D, CompleteOnResponse>, T>> + Clone
+
+#[derive(Clone, Debug)]
+pub struct EwmaConfig {
+    pub default_rtt: Duration,
+    pub decay: Duration,
+}
+
+/// Configures a stack to resolve `T` typed targets to balance requests over
+/// `M`-typed endpoint stacks.
+#[derive(Debug)]
```

Comment:
nit, may not be important: derived `Debug` impl requires that `I: Debug`, which it shouldn't actually need...